File: backend/src/controllers/usersController.js
const bcrypt = require('bcryptjs');
const { validationResult } = require('express-validator');
const User = require('../models/User');

const getUsers = async (req, res, next) => {
  try {
    const { role_id, branch_id, search, is_active, page = 1, limit = 20 } = req.query;

    const filters = {};
    if (role_id) filters.role_id = parseInt(role_id);
    if (branch_id) filters.branch_id = parseInt(branch_id);
    if (search) filters.search = search;
    if (is_active !== undefined) filters.is_active = parseInt(is_active);

    const pagination = { page: parseInt(page), limit: Math.min(parseInt(limit), 100) };
    const { data, total } = await User.findAll(filters, pagination);

    return res.status(200).json({
      success: true,
      data,
      pagination: {
        page: pagination.page,
        limit: pagination.limit,
        total,
        totalPages: Math.ceil(total / pagination.limit)
      }
    });
  } catch (err) {
    next(err);
  }
};

const createUser = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const { name, email, password, role_id, branch_id, is_active } = req.body;

    const existing = await User.findByEmail(email);
    if (existing) {
      return res.status(409).json({ success: false, message: 'Email already in use' });
    }

    const hashed = await bcrypt.hash(password, 10);
    const user = await User.create({ name, email, password: hashed, role_id, branch_id, is_active });

    res.locals.createdId = user.user_id;
    const { password: _pw, ...safeUser } = user;
    return res.status(201).json({ success: true, data: safeUser });
  } catch (err) {
    next(err);
  }
};

const getUserById = async (req, res, next) => {
  try {
    const user = await User.findById(parseInt(req.params.id));
    if (!user) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }
    const { password: _pw, ...safeUser } = user;
    return res.status(200).json({ success: true, data: safeUser });
  } catch (err) {
    next(err);
  }
};

const updateUser = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const id = parseInt(req.params.id);
    const existing = await User.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    const { name, email, role_id, branch_id, is_active } = req.body;
    const updated = await User.update(id, { name, email, role_id, branch_id, is_active });

    const { password: _pw, ...safeUser } = updated;
    return res.status(200).json({ success: true, data: safeUser });
  } catch (err) {
    next(err);
  }
};

const deleteUser = async (req, res, next) => {
  try {
    const id = parseInt(req.params.id);
    const existing = await User.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    if (id === req.user.user_id) {
      return res.status(400).json({ success: false, message: 'Cannot delete your own account' });
    }

    await User.update(id, { is_active: 0 });
    return res.status(200).json({ success: true, message: 'User deactivated successfully' });
  } catch (err) {
    next(err);
  }
};

module.exports = { getUsers, createUser, getUserById, updateUser, deleteUser };

File: dotnet-api/Controllers/UsersController.cs
using ActivityTrackerAPI.Helpers;
using ActivityTrackerAPI.Models.DTOs;
using ActivityTrackerAPI.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ActivityTrackerAPI.Controllers;

/// <summary>User management endpoints</summary>
[ApiController]
[Route("api/users")]
[Authorize]
[Produces("application/json")]
public class UsersController : ControllerBase
{
    private readonly IUserService _userService;

    public UsersController(IUserService userService)
    {
        _userService = userService;
    }

    /// <summary>List users with optional filters and pagination</summary>
    [HttpGet]
    [ProducesResponseType(200)]
    public async Task<IActionResult> GetUsers(
        [FromQuery] uint? role_id,
        [FromQuery] uint? branch_id,
        [FromQuery] bool? is_active,
        [FromQuery] string? search,
        [FromQuery] int page = 1,
        [FromQuery] int limit = 20)
    {
        var roleName = User.GetRoleName();
        var currentBranchId = User.GetBranchId();

        // Branch Manager / Team Leader can only see users in their branch
        if ((roleName == "Branch Manager" || roleName == "Team Leader") && branch_id == null)
            branch_id = currentBranchId;

        limit = Math.Min(limit, 100);
        var (data, total) = await _userService.GetAllAsync(role_id, branch_id, is_active, search, page, limit);

        return Ok(new
        {
            success = true,
            data,
            pagination = new { page, limit, total, totalPages = (int)Math.Ceiling((double)total / limit) }
        });
    }

    /// <summary>Create a new user (Admin only)</summary>
    [HttpPost]
    [ProducesResponseType(201)]
    [ProducesResponseType(400)]
    public async Task<IActionResult> CreateUser([FromBody] CreateUserRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var roleName = User.GetRoleName();
        if (roleName != "Admin")
            return StatusCode(403, new { success = false, message = "Access denied. Required roles: Admin" });

        var hashedPassword = BCrypt.Net.BCrypt.HashPassword(request.Password);
        var user = await _userService.CreateAsync(
            request.Name, request.Email, hashedPassword,
            request.RoleId, request.BranchId, request.IsActive);

        return StatusCode(201, new { success = true, data = user });
    }

    /// <summary>Get a user by ID</summary>
    [HttpGet("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> GetUser(uint id)
    {
        var user = await _userService.GetByIdAsync(id);
        if (user == null)
            return NotFound(new { success = false, message = "User not found" });

        return Ok(new { success = true, data = user });
    }

    /// <summary>Update a user (Admin only)</summary>
    [HttpPut("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(403)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> UpdateUser(uint id, [FromBody] UpdateUserRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var roleName = User.GetRoleName();
        if (roleName != "Admin")
            return StatusCode(403, new { success = false, message = "Access denied. Required roles: Admin" });

        var existing = await _userService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "User not found" });

        var updated = await _userService.UpdateAsync(id, request.Name, request.Email,
            request.RoleId, request.BranchId, request.IsActive);

        return Ok(new { success = true, data = updated });
    }

    /// <summary>Soft-delete (deactivate) a user (Admin only)</summary>
    [HttpDelete("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(403)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> DeleteUser(uint id)
    {
        var roleName = User.GetRoleName();
        if (roleName != "Admin")
            return StatusCode(403, new { success = false, message = "Access denied. Required roles: Admin" });

        var existing = await _userService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "User not found" });

        await _userService.UpdateAsync(id, null, null, null, null, false);
        return Ok(new { success = true, message = "User deactivated successfully" });
    }
}
